Settings UI: update Akismet settings card with a field to enter API key

eliorivero opened the pull request:
Saving/fetching data requires #6693 to be merged. This one handles the UI.

Fixes #6206
Fixes #6207

<img width="748" alt="captura de pantalla 2017-03-21 a las 16 31 35" src="https://cloud.githubusercontent.com/assets/1041600/24166976/e83f30a8-0e53-11e7-8916-7fdcea7bb629.png">

<img width="746" alt="captura de pantalla 2017-03-21 a las 16 26 22" src="https://cloud.githubusercontent.com/assets/1041600/24166984/eb45b542-0e53-11e7-94dc-9d4eea92ab25.png">

<img width="744" alt="captura de pantalla 2017-03-21 a las 16 27 13" src="https://cloud.githubusercontent.com/assets/1041600/24166988/edd7b864-0e53-11e7-9e93-683f0209b4f8.png">

<img width="818" alt="captura de pantalla 2017-03-21 a las 16 27 01" src="https://cloud.githubusercontent.com/assets/1041600/24166996/f22f6d6c-0e53-11e7-82e0-cf019c1845b6.png">

<img width="756" alt="captura de pantalla 2017-03-21 a las 16 27 06" src="https://cloud.githubusercontent.com/assets/1041600/24166999/f4a4c3d0-0e53-11e7-8025-5680b6bc236d.png">

#### Changes proposed in this Pull Request:

* card is hidden if Akismet is inactive
* gone is the checkbox for number of approved comments
* introduces input field to add the API key
* balances the margins of SettingsGroup when it's inside FoldableCard

#### Testing instructions:

* You should see the API key field. If you rebase `add/akismet-save-key`

georgestephanis commented on _inc/client/security/antispam.jsx:
```diff
@@ -3,36 +3,54 @@
+										disabled={ this.props.isSavingAnyOption( 'wordpress_api_key' ) }
+										onChange={ this.props.onOptionChange } />
+								</FormLabel>
+								<p className="jp-form-setting-explanation" >
```
Would it be a good idea to maybe hide this if they already have a valid api key?

dereksmart commented:
This is what we're doing on the dashboard with an invalid key (no key)

Maybe we just do something like this for the card? 

![screen shot 2017-03-20 at 10 12 16 am](https://cloud.githubusercontent.com/assets/7129409/24103482/cc68daa4-0d55-11e7-86c5-bf70d108f853.png)

eliorivero commented:
Regarding the text, I followed the layout and text that @ashleighaxios proposed for Calypso https://github.com/Automattic/wp-calypso/issues/12195
While I do agree that it's a bit confusing, for this pass we can go with it and tweak it later.

Nice catch about removing the key, and yeah, when the key is invalid, we should do something. Let's do the same than the dash item card you pointed.

eliorivero commented:
> Users that don't know what an API Key is and don't really need to know -- it overcomplicates things for them.

Not sure if this is a concern for now. The field is hidden in a foldable card, so it's hidden unless user opens it. Even if they open it, to edit it, they still have to intentionally enter text and click the saving button.

However, one thing you mentioned 
> "Your API key is valid!"
makes me think if we should add something, like a checkmark ✅, that provides feedback to user that the current key in the field is valid. What do you think @ashleighaxios?

dereksmart commented:
> makes me think if we should add something, like a checkmark

Maybe we use a compact green notice like we do for backups? 
![screen shot 2017-03-20 at 10 31 14 am](https://cloud.githubusercontent.com/assets/7129409/24104360/6b14064a-0d58-11e7-8c00-45310538f2e6.png)

eliorivero commented:
Nice, that would be consistent with the security scanning card.

georgestephanis commented:
Ah, I missed that it was a foldable card.

I'd still like the blurb afterwards to be contextually relevant to the current api key -- cc: @michelleweber or anyone from editorial for suggestions / direction?

jeffgolenski commented:
@eliorivero per your request, and with the gracious help of @ashleighaxios here's a mockup of how this section should appear. 

![screen shot 2017-03-20 at 12 37 28 pm](https://cloud.githubusercontent.com/assets/214813/24110768/b5d34850-0d6a-11e7-8774-0b926f024480.png)

We want to ensure we're matched up with calypso in how we handle form validation messages. Also, let's use the label "antispam API key" (swapping 'antispam' if we're using a different name for the feature, of course) where we can—to prevent any possible API confusion within wp-admin. Sure the form lives in the spam filtering card, but like you said. better safe than sorry!

georgestephanis commented:
Should antispam be capitalized or not?  The mockup and your comment seem to vary in usage.

jeffgolenski commented:
@georgestephanis Fixed! It's a feature name. Might as well capitalize it!

eliorivero commented:
@jeffgolenski just noticed you dropped a link. Where should the "Learn more" go to? Shouldn't it have period at the end?

eliorivero commented:
No worries, will point it to `https://docs.akismet.com/getting-started/api-key/`

eliorivero commented:
Updated:

Displays messages when key is valid and invalid

<img width="757" alt="captura de pantalla 2017-03-20 a las 18 26 14" src="https://cloud.githubusercontent.com/assets/1041600/24122491/c185c2ce-0d9a-11e7-8a47-43dd967e1c32.png">

<img width="707" alt="captura de pantalla 2017-03-20 a las 18 26 55" src="https://cloud.githubusercontent.com/assets/1041600/24122512/d31c2ee2-0d9a-11e7-9155-24f6dd9c6d5f.png">

eliorivero commented:
Updates:

- hide valid/invalid message when text input is cleared
- add new endpoint /module/<service>/key/check to check the validity of an API key for a service. Introduces methods to check Akismet API key.
- add actions/reducers to check the validity of an Akismet API key
- add new `QueryAkismetKeyCheck` component 
- add new endpoint supporting POST method that receives a key to check against a whitelisted service. Introduced methods to check against Akismet.
- allow to pass a key to check against a service on the fly as the user types it.
- when key is first retrieved block input field and show message in it. When key check finishes display a success message if key is valid and an error otherwise. When user types the key, display message that it's being checked and show success or error message afterwards

jeffgolenski commented:
@eliorivero 

Solid work here! I'm curious. If the API key has been inserted and its valid. Can we hide the block of text that starts with "If you don't already have an API key get on here..."?

eliorivero commented:
Sure @jeffgolenski 👍

eliorivero commented:
Updated:

- only show explanation message under input field when key is invalid or field is empty.

Screenshots in PR description were updated.

ashleighaxios commented:
Funny little block, but you made it work great. Ty

eliorivero commented:
@georgestephanis @dereksmart I'm worried that someone could use the endpoint to guess keys. Should we limit the number of key checks during a minute?
The UI interface has a debounce of half second, but the endpoint doesn't, and maybe we should add some protection. Then again, not sure if I'm being overzealous.

dereksmart commented:
@eliorivero Why not just use a more strict permissions check?  Is there a reason we need to make the $_GET request available to anyone who has access to the admin?  Why not limit to `manage_plugin` capability or something

eliorivero commented:
Updates:

- check that user can `manage_options` (admin role) to use the API check endpoint.

georgestephanis commented:
"guided through the process of getting one in a new window" -- is that overly specific?

Could this be confusing if it's actually a new tab?

Am I overthinking this?

Probably.

dereksmart commented:
@eliorivero think it makes sense to disable the save button when the API key is rendered `invalid`?  Since saving won't work anyway...

![screen shot 2017-03-22 at 10 12 50 am](https://cloud.githubusercontent.com/assets/7129409/24202017/315c330a-0ee8-11e7-8857-45065300abbe.png)

dereksmart commented:
When there is no API key, the UI still says that my site is protected from Spam. 

![screen shot 2017-03-22 at 10 15 19 am](https://cloud.githubusercontent.com/assets/7129409/24202114/85b65d54-0ee8-11e7-984e-db67c4604bc8.png)

dereksmart commented:
> "guided through the process of getting one in a new window" -- is that overly specific?

Well no, considering that currently the link in the card opens in the same window :)

dereksmart commented:
Also, if the site has a paid plan, they simply could get the key re-activated just by going to the setup flow for akismet: `wordpress.com/plugins/setup/dereksmart.wpsandbox.me?only=akismet`

We should be doing this for plan accounts, but take them to the normal flow for those with akismet active without a plan.

eliorivero commented:
Updated:

- don't enable the Save settings button if the key entered is invalid
- update foldable card text based on API key validity
- removed "in a new window" since we open in the same tab

dereksmart commented:
Looks great elio! send the 🐑

eliorivero commented:
@dereksmart I've addressed your remarks

- don't accept any whitespace.
- allow to save an empty key for Akismet key.
- allow checking for a key validity when the current key is invalid

dereksmart commented:
This is working very nicely now.  Thank you @eliorivero for your hard work on this! 

🐑 !!!
